server/src/main/java/com/ncc/neon/services/QueryService.java

```diff
@@ -22,13 +21,23 @@
 
     private QueryAdapterLocator queryAdapterLocator;
 
-    QueryService(QueryAdapterLocator queryExecutorLocator) {
+    private ClusterService clusterService;
+
+    @Autowired
+    QueryService(QueryAdapterLocator queryExecutorLocator, ClusterService clusterService) {
         this.queryAdapterLocator = queryExecutorLocator;
+        this.clusterService = clusterService;
     }
 
     public Mono<TabularQueryResult> executeQuery(ConnectionInfo ci, Query query) {
         QueryAdapter adapter = this.queryAdapterLocator.getAdapter(ci);
-        return adapter.execute(query);
+
+        if (query.getClusterClause() != null) {
+            this.clusterService.setClusterClause(query.getClusterClause());
+            return adapter.execute(query).flatMap(this.clusterService::cluster);
```

Comment:
Out of curiosity, why use `flatMap` here instead of `map`?  I can never remember the difference.

Reply:
I read that if you want to do things async you use `flatMap`. The other thing is if you know exactly how long it'll take to do the operation, you do `map`. I can change it to `map` if you think that's more appropriate.

Reply:
Awesome!  Sounds good.  Thanks.